sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/aio/_client.py

```diff
@@ -0,0 +1,1091 @@
+            try:
+                await self._ws.connect()
+            except aiohttp.ClientConnectorError as e:
+                raise OpenClientError("Fail to open client") from e
+
+            _LOGGER.debug("WebSocket connection has opened")
+            self._state = WebPubSubClientState.CONNECTED
+            self._task_listen = asyncio.create_task(_listen())
+
+        if self._is_stopping:
+            raise OpenClientError("Can't open a client during closing")
+        self._cancel_task(self._task_seq_ack)
+
+        self._ws = WebSocketAppAsync(
+            url=url,
+            on_open=on_open,
+            subprotocols=[self._protocol.name] if self._protocol else [],
+            header={_USER_AGENT: format_user_agent(self._user_agent)},
+        )
+        await self._ws.run_forever()
+
+        # set coroutine to check sequence id if needed
+        if self._protocol.is_reliable_sub_protocol:
+
+            async def sequence_id_ack_periodically():
+                while self._is_connected():
+                    try:
+                        is_updated, seq_id = self._sequence_id.try_get_sequence_id()
+                        if is_updated and seq_id is not None:
+                            await self._send_message(SequenceAckMessage(sequence_id=seq_id))
+                    finally:
+                        await asyncio.sleep(1.0)
+
+            self._task_seq_ack = asyncio.create_task(sequence_id_ack_periodically())
+
+        _LOGGER.info("connected successfully")
+
+    async def _start_core(self):
+        self._state = WebPubSubClientState.CONNECTING
+        _LOGGER.info("Staring a new connection")
+
+        # Reset before a pure new connection
+        self._sequence_id.reset()
+        self._is_initial_connected = False
+        self._last_close_event = None
+        self._last_disconnected_message = None
+        self._connection_id = None
+        self._reconnection_token = None
+        self._url = None
+
+        self._url = self._credential.get_client_access_url()
+        await self._connect(self._url)
+
+    async def open(self) -> None:
+        """open the client and connect to service"""
+
+        if self._is_stopping:
+            raise OpenClientError("Can't open a client during stopping")
+        if self._state != WebPubSubClientState.STOPPED:
+            raise OpenClientError("Client can be only started when it's Stopped")
+
+        try:
+            await self._start_core()
+        except Exception as e:
+            self._state = WebPubSubClientState.STOPPED
+            self._is_stopping = False
+            raise e
+
+    async def close(self) -> None:
+        """close the client"""
+
+        if self._state == WebPubSubClientState.STOPPED or self._is_stopping:
+            return
+        self._is_stopping = True
+        old_tasks = [self._task_listen, self._task_seq_ack]
+
+        await self._ws.close()  # type: ignore
+        for task in old_tasks:
+            self._cancel_task(task)
```

Comment:
You can use the pattern we use in EH to stop these background tasks [here](https://github.com/Azure/azure-sdk-for-python/blob/d58e28f66e52b0199cdf843e4a9b10f634c348a1/sdk/eventhub/azure-eventhub/azure/eventhub/aio/_transport/_pyamqp_transport_async.py#L250)

We basically use a while loop with a bool variable, when we want to stop it we set it to False and then await the task. This way you done have to cancel them

Reply:
Yeah, it is better to wait for the listen task and sequence ack task complete.